dashboard threat model (#67)
radius-project/design-notes: architecture/2024-08-dashboard-component-threat-model.md

# Radius Dashboard Component Threat Model

- **Author**: @nithyatsu

## Overview

This document provides a threat model for the Radius Dashboard component. It identifies potential security threats to this part of Radius and suggests possible mitigation. The document includes an analysis of the system, its assets, identified threats, and recommended security measures to protect the system.

The Radius Dashboard component provides the frontend experience for Radius. 
It provides visual and textual representation of user's applications, environments and recipes.

## Terms and Definitions

| Term                  | Definition      |
| --------------------- | ----------------------------- |
| UCP | Universal Control Plane for Radius |
| DoS | Denial of Service |
| React Application | A web application built using the React library, which allows for the creation of dynamic and interactive user interfaces through a component-based architecture|

## System Description

> The Dashboard component is an instance of [Backstage](https://backstage.io/). We customize Backstage by installing a Radius plugin and the community-supported Kubernetes plugin. The Dashboard is a client of the Radius API. It queries the graph of an application or a list of environment and constructs a visual representation of the response.

Reviewer: This is very clear now, thank you 🙏

  listen: ':7007'
```

3.  use sqlite DB. This is light weight, not accessible to users directly and contains no useful information.
  
```
app-config.yaml :

backend:
  database:
    client: better-sqlite3
    connection: ':memory:'
```

At present, Dashboard can only present the Radius application metadata visually. It has no ability to Create, Modify, Update or Delete any of the Radius application resources. This eliminates the scope of threats that require "write" action using the Dashboard portal.

#### Storage of secrets

None

#### Data Serialization / Formats

None

### Cryptography 

None

### Clients

The primary user of Dashboard is browser. At present, we dont have any other Backstage plugin that cloud be a Radius Dashboard client but that could change in future. 

## Trust Boundaries

We have a few different trust boundaries for the Dashboard component:

- **Kubernetes Cluster**: The overall environment where the Dashboard component operates and serves clients.
- **Namespaces within the Cluster**: Logical partitions within the cluster to separate and isolate resources and workloads. 

> The Dashboard component lives inside the `radius-system` namespace in the Kubernetes cluster where it is installed. UCP also resides within the same namespace.Namespaces within Kubernetes can help set Role-Based Access Control (RBAC) policies.

Reviewer: ```suggestion
The Dashboard component lives inside the `radius-system` namespace in the Kubernetes cluster where it is installed. UCP also resides within the same namespace. Namespaces within Kubernetes can help set Role-Based Access Control (RBAC) policies.
```

# Radius Dashboard Component Threat Model

- **Author**: @nithyatsu

## Overview

This document provides a threat model for the Radius Dashboard component. It identifies potential security threats to this part of Radius and suggests possible mitigations. The document includes an analysis of the system, its assets, identified threats, and recommended security measures to protect the system.

The Radius Dashboard component provides the frontend experience for Radius. 
It provides visual and textual representation of user's applications, environments and recipes.

## Terms and Definitions

| Term                  | Definition      |
| --------------------- | ----------------------------- |

> | mTLS | Mutual Transport Layer Security (mTLS) allows two parties to authenticate each other during the initial connection of an SSL/TLS handshake. |

Reviewer: Is there any mention of mTLS in this doc? Feels like it might be a copy-paste.

> | UCPD | Universal Control Plane Daemon for Radius |

Reviewer: Do we need to talk about `UCPD` specifically? It feels more accessible to talk about UCP as a system, than `UCPD` (UCP as a process/container).

| DoS | Denial of Service |

> | SPA | Single Page Application (SPA): Typically a React application that contains the logic for most user interactions within the React components. SPAs do not require reloading the page from the server during user interactions. |

Reviewer: This feels a little bit clumsy. SPA != React. There are other SPA frameworks. Might be simpler for readers if we just say what React is and provide a link.

## System Description

> The Dashboard component is an SPA built as [Backstage](https://backstage.io/) plugin. It is a client of Radius API. It queries the graph of an application or a list of environment and constructs a visual representation of the response.

Reviewer: ```suggestion
The Dashboard component is an instance of [Backstage](https://backstage.io/). We customize Backstage by installing a Radius plugin and the community-supported Kubernetes plugin. The Dashboard is a client of the Radius API. It queries the graph of an application or a list of environment and constructs a visual representation of the response.
```

Reviewer: It's not just the plugin. We're shipping users an instance of Backstage.

> ### Architecture

Reviewer 2: Are we thinking that the dashboard is an extension point for customers? If so, does that affect the threat model in any way?

Author: Dashboard does not have extendability.

![Dashboard Architecture](2024-08-dashboard-component-threat-model/dashboard-arch.png)

> Given that the Radius Dashboard is developed as a Backstage plugin, it is essential to first examine the Backstage architecture. Backstage provides a core Single Page Application (SPA), a core backend, and the ability to configure a desired database. The core functionality can be enhanced using plugins.

Reviewer: Can you clean this up? See above ^^^

> To deploy the Radius Dashboard as a Backstage plugin, we have implemented a Radius SPA plugin that integrates with the core Backstage SPA, and a Radius backend plugin that integrates with the core Backstage backend. While the backend can technically be accessed directly, our setup restricts this access. We configured the Radius Dashboard App to use a lightweight, in-memory SQLite database, which stores no data and is not accessible directly from outside.

Reviewer: Can you clean up the terminology here? We're shipping users an instance of Backstage that contains our plugin.

This application is run as dashboard pod in radius-system namespace along with other Radius pods. 

### Implementation Details

> The Radius Dashboard is developed as a Backstage Plugin, making it dependent on the Backstage framework for both display and backend functionality. For detailed information on Backstage's threat model, refer to the [Backstage Threat Model](https://backstage.io/docs/overview/threat-model/).

Reviewer: Please clean up terminology re: instance of backstage vs plugin.

The application uses three config files: app-config.yaml, app-config.local.yaml,

> app-config.dashboard.yaml. 

Reviewer: Is this true? Which of these are shipped to the user.

  listen: ':7007'
```

1.  use sqlite DB. This is light weight, not accessible to users directly and contains no useful information.
  
```
app-config.yaml :

backend:
  database:
    client: better-sqlite3
    connection: ':memory:'
```

At present, Dashboard can only present the Radius application metadata visually. It has no ability to Create, Modify, Update or Delete any of the Radius application resources. This eliminates the scope of threats that require "write" action using the Dashboard portal.

#### Storage of secrets

None

#### Data Serialization / Formats

None

### Cryptography 

None

### Clients

The primary user of Dashboard is browser. At present, we dont have any other Backstage plugin that cloud be a Radius Dashboard client but that could change in future. 

## Trust Boundaries

We have a few different trust boundaries for the Dashboard component:

- **Kubernetes Cluster**: The overall environment where the Dashboard component operates and serves clients.
- **Namespaces within the Cluster**: Logical partitions within the cluster to separate and isolate resources and workloads.

> #### Key Points of Namespaces

Reviewer: Do we need to copy all of these details from the Controller document?

Reviewer: What would happen if we didn't?

Author: Add details on how namespace is helpful in RBAC

1. **Isolation of Resources and Workloads**: Different namespaces separate and isolate resources and workloads within the Kubernetes cluster.
2. **Access Controls and Permissions**: Access controls and other permissions are implemented to manage interactions between namespaces.
3. **Separation of Concerns**: Namespaces support the separation of concerns by allowing different teams or applications to manage their resources independently, reducing the risk of configuration errors and unauthorized changes.

The Dashboard component lives inside the `radius-system` namespace in the Kubernetes cluster where it is installed. UCPD also resides within the same namespace.

The webapp portal is accessible to various configured users. Quoting from Backstage threat model, these users could belong to one of these trust levels:

> **An internal user** is an authenticated user that generally belongs to the organization of a particular Backstage deployment. These users are trusted to the extent that they are not expected to compromise the availability of Backstage, but they are not trusted to not compromise data confidentiality or integrity.

Reviewer: Can you find a way to simplify this for readers? It's not super clear that we're quoting. With all of the test here, it's a lot to read and I think readers will want to give feedback on this.

Author: can add a link and summarize instead.

```
kubectl port-forward --namespace=radius-system svc/dashboard 7007:80
``` 

Decisions to make Dashboard public-facing should be the user's conscious choice. 

## Assumptions

This threat model assumes that:

1. The Radius installation is not tampered with.
2. The Kubernetes cluster that Radius is installed on is not compromised.
3. It is the responsibility of the Kubernetes cluster to authenticate users. Administrators and users with sufficient privileges can perform their required tasks. Radius cannot prevent actions taken by an administrator.
4. Dashboard users have been configured to have right level of access by following [Backstage Threat Model](https://backstage.io/docs/overview/threat-model/). These users are trusted to the extent that they are not expected to compromise the availability of Backstage, but they are not trusted to not compromise data confidentiality or integrity.
5. Dashboard is not public facing.
6. Access to Dashboard is using HTTPS.
7. Authentication mechanism provided by Backstage is robust.

> 

Reviewer: This is a really good set of assumptions and explained well.

## Data Flow

### Diagram

![Radius Dashboard](2024-08-dashboard-component-threat-model/dashboard-tm.png)

1. User types the backstage url and accesses Radius plugin
2. Request reaches the dashboard pod in `radius-system` namespace in kubernetes cluster.
3. The dashboard service sends a Radius API request to UCP.
4. UCP works with ApplicationsCore-RP and sends response back to Dashboard SPA.
5. Dashboard SPA contructs the visuals using components from backstage core, rad-component and data in API response and responds with appropriate page to the user. 

> ### Threats

Reviewer 2: Would it be possible to add a malicious backstage plugin?

 
#### Threat 1: DoS

**Description**

A client can access Dashboard repeatedly or fetch the page in a loop.

**Impact**

Due to the volume of requests Dashboard as well as the UCP, AppCore-RP components involved in serving the request could run out of resource to serve a legitimate request.
   

> **Mitigation**:

Reviewer 3: Have we explored an option to configure rate limiting?

Author: we discussed this in meeting and how we can use permission system to make sure we dont get too many requests (which would be from unauthorized users)

1. Access to Dashboard should be provided to trusted users.  The [Backstage  permissions system](https://backstage.io/docs/permissions/overview) should be enabled and configured to restrict access as necessary.
   
**Status**:

> Active. Operators are expected to configure the system and limit access to Dashboard portal. 

Reviewer 3: Is this something we have provide guidance on in Radius docs?

Author: Not today. We should add docs on how this can be done. (create ticket)

> 2. Audit logs should be enabled to monitor and report on suspicious user activity. 

Reviewer: Is this something backstage provides?

**Status**

> [In Progress](https://github.com/backstage/backstage/issues/23950)

Reviewer 3: Will this be enabled by default or opt in?

Author: Follow up once the feature is implemented and make a decision.

Author: created https://github.com/radius-project/dashboard/issues/105 to address the documentation which captures the remediation/ future revisit.

#### Threat 2: Information Disclosure by unauthorized access to application information

**Description**

Access to app graph through the portal can provide information on dependency between app components.

**Impact**:

A malicious user can utilize the app graph to stage effective attack by targeting a component that has most dependency.

**Mitigation**:

1. Access to Dashboard portal should be provided to trusted users. While we dont expose any secrets in db, users should still enable authentication and secure access to data based on roles.
   
**Status**:

Active. Operators are expected to configure the [Backstage  permissions system](https://backstage.io/docs/permissions/overview) to restrict access as necessary. 

2. Audit logs should be enabled to monitor and report on suspicious user activity. 

**Status**

[In Progress](https://github.com/backstage/backstage/issues/23950)

#### Threat 3: Spoofing dashboard service-account can cause DoS 
**Description**

> If an unauthorized user or malicious actor gtampers with cluster,

Reviewer 3: nit: typo

```suggestion
If an unauthorized user or malicious actor tampers with cluster,
```

he can create services with dashboard service-account to send Radius API requests. 

**Impact**:

A malicious user can utilize Radius API and create too many resources. 

> This is because dashboard service-account is allowed get, list, and post requests to UCP

Reviewer 3: Would be helpful to set some context in the architecture section on why `post` permission is needed.

If HTTPS is not enabled and the dashboard is made public-facing then a malicious user would be able to see the data used to construct dashboard by sniffing.

**Impact**:

If an unauthorized user or malicious actor gathers data used to construct the page by viewing the network traffic, he can get access to information about application and use it to stage attacks. 

**Mitigation**:

1. Dashboard is not intended to be public-facing. If a decision is taken to make it public, it should be configured to use HTTPS and The [Backstage  permissions system](https://backstage.io/docs/permissions/overview) should be enabled and configured to restrict access as necessary.
   
**Status**:

Active. Operators are expected to limit access to Dashboard portal. Further, 
they should configure Dashboard to use HTTPS if it is public-facing.

## Open Questions

## Action Items

1. Dashboard should be accessed only on HTTPS if it should be available outside cluster. Currently, we can access the application on http but since we only access the application on localhost using =kubernetes port-forward, this is OK. 
 

> 2. Enable authentication on Dashboard. This could be tied to RBAC support on Radius, since we might want the same users to be allowed dashboard logins by default with permissions configured using Backstage permission system.

Reviewer: What's the actual action item? Is this something the user needs to do? If that's the case what's the action item for us?

Author: Add documentation in dashboard related info to capture this.

Author: create a story
